### PR TITLE
Real fix for collection creation bug

### PR DESCRIPTION
**Pull request by noblepaul**

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java [hunk 1]
```diff
@@ -345,6 +346,14 @@ private ZkWriteCommand updateState(final ClusterState prevState, ZkNodeProps mes
     Map<String, Replica> replicas;
 
     if (slice != null) {
+      if (coreNodeName != null) {
+        if (isModifiedReplica(slice.getReplica(coreNodeName), message)) {
+          //one of the basic properties of the replica got changed such ar core/base_url/ node_name
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:351]: When/how core name can change?

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 2]
```diff
@@ -1703,7 +1703,8 @@ public void publish(final CoreDescriptor cd, final Replica.State state, boolean
       DocCollection coll = zkStateReader.getCollection(collection);
       if (forcePublish || sendToOverseer(coll, coreNodeName)) {
         overseerJobQueue.offer(Utils.toJSON(m));
-      } else {
+      }
+      if (coll != null && coll.isPerReplicaState() && coreNodeName != null) {
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: I feel we should wait here to update state.json file?

Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: @noblepaul / @chatman any thought on this!

File: solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java (hunk 1 above)
Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:351]: CCing @noblepaul @chatman

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java (hunk 2 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: Why? Why should it wait?

File: solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java (hunk 1 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:351]: This is actually only pertinent to our JUnit tests. When we restart a node it gets assigned a new port. So, the state message handling in overseer is written to deal with it

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java (hunk 2 above)
Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: 1. nonprs collection updates state.json atomically.
2. With prs here, we will update synchronously, but there is possibility that state.json will be updated later.

File: solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java (hunk 1 above)
Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:351]: port is ok but why corename?

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:351]: Overseer can handle a change in core name as well. So, I  just defensively added that check

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java (hunk 2 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: Even if it's not refreshed atomically, it should be just fine

Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: In PRS, we update structure of collection first in stats.json, and then update the prs state. We have written go-client with such assumption. I don't think we want to break this assumption of prs protocol.

File: solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java (hunk 1 above)
Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:351]: I won't do this for PRS.

Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:351]: This way we don't need to update PRS state - just update the state.json file for port/http/s change

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java (hunk 2 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: The data that's updated by the data node (prs states) is not the same as the data updated by overseer (state.json). Hence, they can be refreshed independently. Also, we should never wait for overseer to execute the state messages. It's horrible for performance
We should write our go client consistent with how Solr and SolrJ behaves and not the other way around

Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: Possibly, data updated in this case is different. But we should keep behavior similar to what non-prs have in this case. Otherwise we might run in some race-conditions, where code is waiting for updates of (shard and replica).

Please let me know for what we have changed this piece of code now?

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: This doesn't matter for us. It's only relevant for the JUnit tests. So, the messages sent to overseer becomes no-op

Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: Are you saying we will never send message to overseer for prs-collection?

I do see during split, it sends message to overseer and then it updates prs states.

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: we send messages to overseer in certain cases. but, they don't result in updating PRS states

Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: We have removed the else part that means there is case where we send two msg one by one?

Also, I observed that during split msg is going to overseer and we are also updating prs state. That's why I'm saying we need to wait here.

Would  be good to write unit test to verify this.

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1707]: > We have removed the else part that means there is case where we send two msg one by one?

We send one mesage to overseer and do one direct update of PRS states. The overseer **may** update some attributes of the replica as a result of the message but it never touches the PRS states

> Also, I observed that during split msg is going to overseer and we are also updating prs state. That's why I'm saying we need to wait here.

We should ensure that overseer never updates PRS states . 

> Would be good to write unit test to verify this.

yes. let's add testcases to ensure this

File: solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java [hunk 3]
```diff
@@ -1430,9 +1430,11 @@ private void refreshAndWatchChildren() throws KeeperException, InterruptedExcept
       Stat stat = new Stat();
       List<String> replicaStates = null;
       try {
-        replicaStates = zkClient.getChildren(collectionPath, this, stat, true);
-        PerReplicaStates newStates = new PerReplicaStates(collectionPath, stat.getCversion(), replicaStates);
         DocCollection oldState = watchedCollectionStates.get(coll);
+        replicaStates = zkClient.getChildren(collectionPath, this, stat, true);
+        PerReplicaStates newStates = PerReplicaStates.fetch (collectionPath, zkClient, oldState == null ?
```
Comment (hiteshk25) [inline solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java]: I just realized that `PerReplicaStates.fetch` method makes calls to zk. And we are already fetching children in previous line here `replicaStates = zkClient.getChildren(collectionPath, this, stat, true);`

Comment (noblepaul) [reply solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java]: true. good catch

File: solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java [hunk 4]
```diff
@@ -252,8 +254,7 @@ private ZkWriteCommand updateState(final ClusterState prevState, ZkNodeProps mes
       log.info("Failed to update state because the replica does not exist, {}", message);
       return ZkStateWriter.NO_OP;
     }
-    boolean persistCollectionState = collection != null && collection.isPerReplicaState();
-
+    boolean persistPrsCollectionState = false;
```
Comment (patsonluk) [inline solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:257]: I personally found this `persistPrsCollectionState` flag quite hard to follow:
1. Many branches and flipping in various places
2. Confusing to dev debugging code with their non PRS collection (the devs might not even be aware of what PRS is) as this flag will be set to `true` in various cases.

By looking at the original design of this method `updateState` - the behavior is well defined - if someone invokes this method, it will generate a single write to `state.json` (ie return `new ZkWriteCommand(collectionName, newCollection);`)

However, we are making special cases for PRS here, and I am wondering if it's truly necessary - if it's not an issue for non PRS collection, perhaps it's not really an issue for PRS too? This should not raise any race condition (as it's writing to state.json only, no longer would it write PRS entries 👍🏼 ) unless data node still has code that updates state.json too (then it's a race condition on state.json, not PRS entries)? There were some places that did ZKStateWriter.setData directly for PRS collection and I think we should fix those instead of making special cases here 😅

File: solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java [hunk 5]
```diff
@@ -321,6 +319,8 @@ public void call(ClusterState clusterState, ZkNodeProps message, @SuppressWarnin
         }
       }
       if(isPRS) {
+        byte[] data = Utils.toJSON(Collections.singletonMap(collectionName, clusterState.getCollection(collectionName)));
+        ocmh.zkStateReader.getZkClient().setData(collectionPath, data, true);
```
Comment (patsonluk) [inline solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java:323]: Is there a good reason that we don't update state.json using `ocmh.overseer.offerStateUpdate` (which is single threaded) ? Doing directly writes here opens up possibility of race condition - even if it does not happen right now, it will always be a liability - as it introduces multiple ways to update state.json

`ocmh.overseer.offerStateUpdate` enforces ordering, load throttling, centralized logging and most importantly single threading to eliminate any race conditions on `state.json` updates. (uberseer could still contend for state.json update, but that's outside of this scope)

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java:323]: > Is there a good reason that we don't update state.json using `ocmh.overseer.offerStateUpdate`

Yes. `ocmh.overseer.offerStateUpdate` is a command and we should be using it only for well known commands. Here we are trying to update the whole state in one go

File: solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java (hunk 4 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:257]: This can be simplified. let me do it anyway

**Comment by noblepaul** [comment]
#172 and #173 fixes this
